crates/ingress-kafka/src/consumer_task.rs

```diff
@@ -457,36 +327,269 @@ impl Drop for AbortOnDrop {
+
+                    ingress_request_counter.increment(1);
+                    trace!(
+                        topic=%self.topic_partition.0, kafka_partition=%self.topic_partition.1,
+                        offset=%offset, consumer_group=%self.consumer_group_id,
+                        "store kafka offset",
+                    );
+                    self.consumer.store_offset(&self.topic_partition.0, self.topic_partition.1, offset)?;
+                },
+                Some(received) = consumer_stream.next() => {
+                    let msg = received?;
+                    let offset = msg.offset();
+                    if legacy_dedup_offset.is_some_and(|dedup_offset| offset as u64 <= dedup_offset) {
+                        // skip duplicated messages. Any gap should be small.
+                        debug!(
+                            topic=%self.topic_partition.0, kafka_partition=%self.topic_partition.1,
+                            offset=%offset, consumer_group=%self.consumer_group_id,
+                            "skipping kafka message (dedup)"
+                        );
+                        self.consumer.store_offset(&self.topic_partition.0, self.topic_partition.1, offset)?;
+                        continue;
+                    }
+
+                    trace!(
+                        topic=%self.topic_partition.0, kafka_partition=%self.topic_partition.1,
+                        offset=%offset, consumer_group=%self.consumer_group_id,
+                        "ingesting kafka message"
+                    );
+
+
+                    let envelope = self.builder.build(producer_id, &self.consumer_group_id, msg)?;
+
+                    // while trying to send the message, we also make sure to process committed messages
+                    // to free up resources asap.
+                    let mut ingest = self.ingestion.ingest(envelope.partition_key(), envelope);
+
+                    let commit_token = loop {
+                        tokio::select!{
+                            biased;
+                            token = &mut ingest => {
+                                break token?;
+                            }
+                            Some(committed) = Self::head_committed(&mut inflight) => {
+                                _ = inflight.pop_front().expect("to exist");
+                                let offset = committed.map_err(|_| Error::IngestionClosed)?;
+
+                                ingress_request_counter.increment(1);
+                                trace!(
+                                    topic=%self.topic_partition.0, kafka_partition=%self.topic_partition.1, offset=%offset,
+                                    "store kafka offset",
+                                );
+                                self.consumer.store_offset(&self.topic_partition.0, self.topic_partition.1, offset)?;
+                            }
+                        }
+                    };
```

Comment:
Not a huge fan of nested tokio selects. Is this strictly required? If it is strictly required, could we move awaiting the commit token to a TaskCenter task and not awaiting it as part of this loop?